File: src/LocalStoreManagement.Desktop/Infrastructure/AppLog.cs
using System.Text;
using Microsoft.Data.Sqlite;

namespace LocalStoreManagement.Desktop.Infrastructure;

public static class AppLog
{
    private static readonly object SyncRoot = new();

    public static string LogDirectory => Path.Combine(AppPaths.DataDirectory, "Logs");
    public static string LogPath => Path.Combine(LogDirectory, "application.log");

    public static void Info(string context, string message)
        => Write(context, message, null);

    public static void Error(string context, Exception exception)
        => Write(context, exception.Message, exception);

    private static void Write(string context, string message, Exception? exception)
    {
        try
        {
            lock (SyncRoot)
            {
                Directory.CreateDirectory(LogDirectory);

                var builder = new StringBuilder();
                builder.Append('[')
                    .Append(DateTimeOffset.Now.ToString("O"))
                    .Append("] ")
                    .Append(context)
                    .AppendLine();
                builder.Append("Database: ").AppendLine(AppPaths.DatabasePath);
                builder.Append("Messaggio: ").AppendLine(message);

                if (exception is SqliteException sqliteException)
                {
                    builder.Append("SQLite code: ").Append(sqliteException.SqliteErrorCode)
                        .Append(" / extended: ").AppendLine(sqliteException.SqliteExtendedErrorCode.ToString());
                }

                if (exception is not null)
                {
                    builder.AppendLine(exception.ToString());
                }

                builder.AppendLine(new string('-', 80));
                File.AppendAllText(LogPath, builder.ToString());
            }
        }
        catch
        {
            // Il logging non deve mai causare a sua volta la chiusura dell'applicazione.
        }
    }
}

File: lib/core/app_log.dart
import 'dart:io';

import 'app_paths.dart';

class AppLog {
  static String get logPath {
    try {
      return '${AppPaths.logsDirectory}${Platform.pathSeparator}application.log';
    } catch (_) {
      final fallback = Directory('${Directory.systemTemp.path}${Platform.pathSeparator}LocalStoreManagementSystem');
      fallback.createSync(recursive: true);
      return '${fallback.path}${Platform.pathSeparator}application.log';
    }
  }

  static void info(String context, String message) => _write(context, message, null, null);

  static void error(String context, Object error, [StackTrace? stackTrace]) =>
      _write(context, error.toString(), error, stackTrace);

  static void _write(String context, String message, Object? error, StackTrace? stackTrace) {
    try {
      final file = File(logPath);
      file.parent.createSync(recursive: true);
      String database;
      try {
        database = AppPaths.databasePath;
      } catch (_) {
        database = '(percorso database non ancora inizializzato)';
      }
      final buffer = StringBuffer()
        ..writeln('[${DateTime.now().toIso8601String()}] $context')
        ..writeln('Database: $database')
        ..writeln('Messaggio: $message');
      if (error != null) buffer.writeln('Errore: $error');
      if (stackTrace != null) buffer.writeln(stackTrace);
      buffer.writeln(List.filled(80, '-').join());
      file.writeAsStringSync(buffer.toString(), mode: FileMode.append, flush: true);
    } catch (_) {
      // Il logging non deve mai causare a sua volta la chiusura dell'applicazione.
    }
  }
}
